graph_algorithms/algorithms/examples/louvain.gsql

```diff
@@ -1,80 +1,295 @@
+                                               deltaQ_new = weightToC - 2 * incident * s.@weight/ @@totalWeight,
+                                               IF deltaQ_new > maxDeltaQ OR (abs(deltaQ_new - maxDeltaQ) < epsilon AND cluster < bestCluster) THEN   # when deltaQ_new is equal to maxDeltaQ, and the cluster label is smaller, also change 
+                                                       maxDeltaQ = deltaQ_new,
+                                                       bestCluster = cluster
+                                               END
+                                       END,
+                                       IF s.@cid != bestCluster THEN 
+                                               @@totIncidentCluster += (s.@cid -> (-1 * s.@weight)),
+                                               @@totIncidentCluster += (bestCluster -> s.@weight),
+                                               s.@cid = bestCluster
+                                       END,
+                                       s.@weightToCluster.clear();
+                        loop = loop + 1;
+                END;
+                last_modularity = @@modularity;
+                @@modularity = 0;
+                T1 = SELECT s
+                     FROM Start:s-(Friend:e)-:t
+                     WHERE s.@cid == t.@cid
+                     ACCUM @@modularity += e.weight - s.@weight * t.@weight / (@@totalWeight);
+                @@modularity = @@modularity / @@totalWeight;                      
+                PRINT iteration AS Phase1Iter, @@modularity;
+                log(debug > 0, "[redrain]#move", iteration, @@modularity);
         END;
 
-        Start = {ANY};
+# Phase 2 --  Merge     
+        iteration2 = 0;
+        WHILE (iteration2 < 2 OR @@modularity2 - last_modularity2 > epsilon) LIMIT iter2 DO
+                iteration2 += 1;
+                Start = SELECT s
+                        FROM Start:s
+                        ACCUM s.@uid = s.@cid;      
+                # Select the vertices with minimal internal id to represent the coarsened graph
+                Start = SELECT s
+                        FROM Start:s 
+                        ACCUM @@representMap += (s.@cid -> s);
+        
+                FOREACH (key, value) IN @@representMap DO
+                        @@representSet += value;                       
+                END;      
+                represent = {@@representSet};
+                @@representMap.clear();
+                @@representSet.clear();
+                log(debug > 0, "[redrain]#2_merge", represent.size()); #@@clusterSizes.size());
 
+            # Get @cweight from totalIncident
+                represent = SELECT s
+                            FROM represent:s
+                            ACCUM s.@cweight = @@totIncidentCluster.get(s.@uid),
+                                  @@clusterSizes += (s.@cid -> 1);
```

Comment:
if represent set is populated based on the existing clusters, will cluster size of each cluster always be 1?

Reply:
Initially the cluster size are all 1, but it changes in the moving phase